pkg/webhook/node/plugins/resourceamplification/resource_amplification.go

```diff
@@ -0,0 +1,162 @@
+}
+
+// Admit makes an admission decision based on the request attributes
+func (n *NodeResourceAmplificationPlugin) Admit(ctx context.Context, req admission.Request, node, oldNode *corev1.Node) error {
+	klog.V(3).Infof("enter NodeResourceAmplificationPlugin plugin admit for %s", req.AdmissionRequest.Name)
+
+	op := req.AdmissionRequest.Operation
+	switch op {
+	case admissionv1.Create:
+		return nil
+	case admissionv1.Update:
+		return n.handleUpdate(oldNode, node)
+	}
+	return nil
+}
+
+// isSupportedResourceChanged checks if the supported resources are changed
+func (n *NodeResourceAmplificationPlugin) isSupportedResourceChanged(oldNode, node *corev1.Node) bool {
+	if oldNode == nil || node == nil {
+		return false
+	}
+	oldOriginalResources := corev1.ResourceList{}
+	newOriginalResources := corev1.ResourceList{}
+	for _, resourceName := range supportedResources {
+		oldResourceValue, ok := oldNode.Status.Allocatable[resourceName]
+		if ok {
+			oldOriginalResources[resourceName] = oldResourceValue
+		}
+		resourceValue, ok := node.Status.Allocatable[resourceName]
+		if ok {
+			newOriginalResources[resourceName] = resourceValue
+		}
+	}
+	return !util.IsResourceListEqual(oldOriginalResources, newOriginalResources)
+}
+
+func (n *NodeResourceAmplificationPlugin) handleUpdate(oldNode, node *corev1.Node) error {
+	if node.Annotations[extension.AnnotationNodeResourceAmplificationRatio] == "" {
+		// When the feature is turned off, the saved raw allocatable needs to be cleaned up
+		delete(node.Annotations, extension.AnnotationNodeRawAllocatable)
+		return nil
+	}
+	if node.Status.Allocatable == nil {
+		// do nothing
+		return nil
+	}
+
+	// Set original resource under two conditions
+	// 1. if the original resource is not set
+	// 2. Node allocatable are updated, such as the administrator adjusting the reserved resource
+	//    https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources
+	//    Here, we rely on an assumption that only kubelet will update allocatable, and other roles will only expand new fields and not overwrite native fields.
+	//    FIXME If the updated resource of kubelet happens to be the same as the amplificated allocatable, the update will not take effect.
+	originalResources := corev1.ResourceList{}
+	if !extension.HasNodeRawAllocatable(node.Annotations) || n.isSupportedResourceChanged(oldNode, node) {
+		// resources in other dimensions will be dynamically updated, so they are not saved here. The caller needs to be aware of this feature.
+		for _, resourceName := range supportedResources {
+			resourceValue, ok := node.Status.Allocatable[resourceName]
+			if ok {
+				originalResources[resourceName] = resourceValue
+			}
+		}
+		if len(originalResources) > 0 {
+			klog.V(2).Infof("admit for %s, set raw allocatable %v", node.Name, originalResources)
+			extension.SetNodeRawAllocatable(node, originalResources)
+		}
+	} else {
+		originalResources, err := extension.GetNodeRawAllocatable(node.Annotations)
```

Comment:
Shall we return this err?

Reply:
Now that we are not returning an error on the webhook, perhaps it would be more appropriate to return an error but configure the webhook to ignore policy.

Reply:
Yeah the current policy of this webhook is ignore, so it's safe to return error here? Returning error also helps if users want to make the policy fail to ensure amplification working, this can make user find issues (misconfiguring, etc.) easier.